### PR TITLE
Ppl manage get methods

### PR DESCRIPTION
**Pull request by nchkim**
*Issue #, if available:*

*Description of changes:*
Added `.list_pipelines()` and `.list_executions()` method for greater access to previous pipelines and executions 

*Testing done:*
Added integration tests for both 

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [ ] I have read the [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md) doc
- [ ] I certify that the changes I am introducing will be backword compatible, and I have discussed concerns about this, if any, with the Python SDK team
- [ ] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md#committing-your-change)
- [ ] I have passed the region in to all S3 and STS clients that I've initialized as part of this change.
- [ ] I have updated any necessary documentation, including [READMEs](https://github.com/aws/sagemaker-python-sdk/blob/master/README.rst) and [API docs](https://github.com/aws/sagemaker-python-sdk/tree/master/doc) (if appropriate)

#### Tests

- [ ] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [ ] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes
- [ ] I have checked that my tests are not configured for a specific region or account (if appropriate)
- [ ] I have used [`unique_name_from_base`](https://github.com/aws/sagemaker-python-sdk/blob/master/src/sagemaker/utils.py#L77) to create resource names in integ tests (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/sagemaker/workflow/pipeline.py [hunk 1]
```diff
@@ -394,6 +426,45 @@ def get_last_execution(self, successful: bool = False, pipeline_arn: str = None)
...
+                    sagemaker_session=self.sagemaker_session,
+                    pipeline=self,
+                )
+            )
+
+        listResponse["PipelineExecutionList"] = pipelineExecutionList
+        if nextToken is not None:
+            listResponse["NextToken"] = nextToken
+
+        return listResponse
```
Comment (qidewenwhen) [inline src/sagemaker/workflow/pipeline.py]: IIRC, Zhankui (@navaj0) has requested to return an object supporting pagination. Can you check with him if the dict return type works?

If no, let's create 2 sibling classes for `PipelineList` and `ExecutionList`

File: tests/integ/sagemaker/workflow/test_workflow.py [hunk 2]
```diff
@@ -1222,17 +1225,103 @@ def test_get_last_execution(sagemaker_session, role, pipeline_name, region_name)
 
         output = pipeline.get_last_execution()
         output_successful = pipeline.get_last_execution(successful=True)
+        list_response = pipeline.list_executions()
+        output_list = list_response["PipelineExecutionList"]
 
-        assert output.arn == execution_fail.arn
-        assert output.pipeline == pipeline
+        equal_pipeline_executions(output, execution_fail)
```
Comment (qidewenwhen) [inline tests/integ/sagemaker/workflow/test_workflow.py]: Nit: `assert_pipeline_executions`

File: tests/unit/sagemaker/workflow/helpers.py [hunk 3]
```diff
@@ -55,6 +55,11 @@ def equal_adjacency_list_with_edges(output, expected):
             assert output_outBoundEdges[j] == expected_outBoundEdges[j]
 
 
+def equal_pipeline_executions(output, expected):
```
Comment (qidewenwhen) [inline tests/unit/sagemaker/workflow/helpers.py]: Seems it's only used in integ test. Let's create a `helpers.py` file under `tests/integ/sagemaker/workflow` and move the helper method there.

File: tests/integ/sagemaker/workflow/test_workflow.py [hunk 4]
```diff
@@ -1222,17 +1225,103 @@ def test_get_last_execution(sagemaker_session, role, pipeline_name, region_name)
...
+        assert pipeline_list[1].parameters == pipeline1.parameters
+        assert len(pipeline_list[1].steps) == 0
+        assert isinstance(pipeline_list[1], ImmutablePipeline)
+
+    finally:
+        try:
+            pipeline1.delete()
+            pipeline2.delete()
+        except Exception:
+            pass
```
Comment (qidewenwhen) [inline tests/integ/sagemaker/workflow/test_workflow.py]: Please run the `black-format`, `pylint`, `flake8`. One new line at the end of file is required

File: tests/integ/sagemaker/workflow/test_workflow.py [hunk 5]
```diff
@@ -1222,17 +1225,103 @@ def test_get_last_execution(sagemaker_session, role, pipeline_name, region_name)
...
+        equal_pipeline_executions(output_list[1], output_successful)
+        assert output_list[1].describe()["PipelineExecutionStatus"] == "Succeeded"
+        assert "NextToken" not in list_response
+
     finally:
         try:
             pipeline.delete()
         except Exception:
             pass
+
```
Comment (qidewenwhen) [inline tests/integ/sagemaker/workflow/test_workflow.py]: Can we also test the case when NextToken is not None? As list executions and list pipelines are quite similar, we can test pagination for list executions only.
We can set the page size to be a small value like 2 and have 3 executions.

This would require us to expose the MaxResults (have sent model link to you in slack)

File: src/sagemaker/workflow/pipeline.py [hunk 6]
```diff
@@ -744,6 +738,20 @@ def upsert(
         raise RuntimeError("Immutable Pipelines cannot be updated")
 
 
+@attr.s
+class PipelineList:
```
Comment (qidewenwhen) [inline src/sagemaker/workflow/pipeline.py]: I guess users won't initiate an object for it directly. Let's make it an internal class: `_PipelineList`

File: src/sagemaker/workflow/pipeline.py [hunk 7]
```diff
@@ -868,6 +876,20 @@ def wait(self, delay=30, max_attempts=60):
         waiter.wait(PipelineExecutionArn=self.arn)
 
 
+@attr.s
+class ExecutionList:
```
Comment (qidewenwhen) [inline src/sagemaker/workflow/pipeline.py]: `_ExecutionList`

File: src/sagemaker/workflow/pipeline.py [hunk 8]
```diff
@@ -744,6 +738,20 @@ def upsert(
...
+class PipelineList:
+    """PipelineList class to encapsulate a list of Pipeline objects
+
+    Attributes:
+        pipeline_list (List[Pipeline]): A list of Pipeline objects
+        next_token (str): If the result of the previous call was truncated, a token that can be used to retrieve
+            the next set of Pipeline objects
+    """
+
+    pipeline_list: Sequence[Pipeline] = attr.ib(factory=list)
```
Comment (qidewenwhen) [inline src/sagemaker/workflow/pipeline.py]: As the step class name is _PipelineList, let's make this attribute name to be pipelines

It's similar to StepCollection and step: https://github.com/aws/sagemaker-python-sdk/blob/master/src/sagemaker/workflow/step_collections.py#L39

File: src/sagemaker/workflow/pipeline.py [hunk 9]
```diff
@@ -868,6 +876,20 @@ def wait(self, delay=30, max_attempts=60):
         waiter.wait(PipelineExecutionArn=self.arn)
 
 
+@attr.s
+class ExecutionList:
+    """ExecutionList class to encapsulate a list of _PipelineExecution objects
+
+    Attributes:
+        pipeline_execution_list (List[_PipelineExecution]): A list of _PipelineExecution objects
```
Comment (qidewenwhen) [inline src/sagemaker/workflow/pipeline.py]: Same above. `pipeline_executions`

File: tests/integ/sagemaker/workflow/test_workflow.py [hunk 10]
```diff
@@ -1279,49 +1301,55 @@ def test_list_pipelines_method(sagemaker_session, role, pipeline_name, region_na
         else_steps=[step_fail],
     )
 
-    pipeline_name2 = pipeline_name + "2"
-    pipeline2 = Pipeline(
-        name=pipeline_name2,
```
Comment (qidewenwhen) [inline tests/integ/sagemaker/workflow/test_workflow.py]: Nit just simply: `name=f"{pipeline_name}3"`

File: tests/integ/sagemaker/workflow/test_workflow.py [hunk 11]
```diff
@@ -1255,17 +1258,36 @@ def test_list_pipelines_method(sagemaker_session, role, pipeline_name, region_na
...
+        CallbackStep(
+            name="callback-step2",
+            sqs_queue_url="https://sqs.us-east-2.amazonaws.com/123456789012/MyQueue",
+            inputs={"arg1": "foo2"},
+            outputs=[outputParam],
+        )
+    ]
+    pipeline_name2 = pipeline_name + "2"
+    pipeline2 = Pipeline(
+        name=pipeline_name2,
```
Comment (qidewenwhen) [inline tests/integ/sagemaker/workflow/test_workflow.py:1285]: Nit just simply: `name=f"{pipeline_name}2"`

**Review by qidewenwhen — APPROVED** [review]
